File: backend/contact-service/app/schemas/contacts.py
from pydantic import BaseModel, Field


class BlockUserRequest(BaseModel):
    user_id: str = Field(min_length=1)
    reason: str | None = None


class BlockedUserResponse(BaseModel):
    user_id: str
    display_name: str | None = None
    blocked_at: str
    reason: str | None = None


class SendContactRequestBody(BaseModel):
    to_user_id: str = Field(min_length=1)
    from_username: str = ""


class ContactRequestResponse(BaseModel):
    id: str
    from_user_id: str
    to_user_id: str
    status: str
    conversation_id: str | None = None
    created_at: str
    resolved_at: str | None = None


class ContactStatusResponse(BaseModel):
    status: str  # "none" | "pending_sent" | "pending_received" | "contacts"
    request_id: str | None = None
    conversation_id: str | None = None

File: frontend/web/src/api/contacts.ts
import { apiFetch } from "./client";

export interface BlockedUser {
  user_id: string;
  display_name: string | null;
  blocked_at: string;
  reason: string | null;
}

export interface ContactRequest {
  id: string;
  from_user_id: string;
  to_user_id: string;
  status: "pending" | "accepted" | "declined";
  conversation_id: string | null;
  created_at: string;
  resolved_at: string | null;
}

export type ContactStatus = "none" | "pending_sent" | "pending_received" | "contacts";

export interface ContactStatusResponse {
  status: ContactStatus;
  request_id: string | null;
  conversation_id: string | null;
}

export async function listBlockedUsers(): Promise<BlockedUser[]> {
  return apiFetch<BlockedUser[]>("/contacts/blocks");
}

export async function blockUser(userId: string, reason?: string): Promise<BlockedUser> {
  return apiFetch<BlockedUser>("/contacts/blocks", {
    method: "POST",
    body: JSON.stringify({ user_id: userId, reason }),
  });
}

export async function unblockUser(userId: string): Promise<void> {
  await apiFetch(`/contacts/blocks/${encodeURIComponent(userId)}`, { method: "DELETE" });
}

export async function sendContactRequest(
  toUserId: string,
  fromUsername: string = "",
): Promise<ContactRequest> {
  return apiFetch<ContactRequest>("/contacts/requests", {
    method: "POST",
    body: JSON.stringify({ to_user_id: toUserId, from_username: fromUsername }),
  });
}

export async function listIncomingRequests(): Promise<ContactRequest[]> {
  return apiFetch<ContactRequest[]>("/contacts/requests/incoming");
}

export async function listOutgoingRequests(): Promise<ContactRequest[]> {
  return apiFetch<ContactRequest[]>("/contacts/requests/outgoing");
}

export async function acceptContactRequest(requestId: string): Promise<ContactRequest> {
  return apiFetch<ContactRequest>(`/contacts/requests/${encodeURIComponent(requestId)}/accept`, {
    method: "PATCH",
  });
}

export async function declineContactRequest(requestId: string): Promise<ContactRequest> {
  return apiFetch<ContactRequest>(`/contacts/requests/${encodeURIComponent(requestId)}/decline`, {
    method: "PATCH",
  });
}

export async function cancelContactRequest(requestId: string): Promise<ContactRequest> {
  return apiFetch<ContactRequest>(`/contacts/requests/${encodeURIComponent(requestId)}/cancel`, {
    method: "PATCH",
  });
}

export async function getContactStatus(userId: string): Promise<ContactStatusResponse> {
  return apiFetch<ContactStatusResponse>(`/contacts/status/${encodeURIComponent(userId)}`);
}
